Tour of csharp

BillWagner opened the pull request:
Fixes #486

Yes, that means the "features and a glance" was reworked into "a tour of C#".

Updated the metadata for these articles.

Note that much of this content came from the introductory section of the C# Language specification. That section is being removed from the official spec. (It is informative, and not normative. The committee decided this amount of informative information did not belong in the spec.) The full text has been reviewed several times as a Word Doc. The Markdown format and all metadata should be reviewed.

BillWagner committed:
first pass outline

BillWagner committed:
interim checkin

BillWagner committed:
interim checkin

BillWagner committed:
grab text from ECMA spec, clause 7.

BillWagner committed:
update with text from C# ECMA spec, clause 7

BillWagner committed:
Update the opening sections.

BillWagner committed:
update the concepts article

BillWagner committed:
update tour of C#

Still more work on that doc, and likely splitting it up

BillWagner committed:
finished first pass review

BillWagner committed:
Review and format

BillWagner committed:
reorganize the tour of C#

BillWagner committed:
Update the Tour of C#

There are different sections for each major topic area.
This makes it much easier to digest in reasonable chunks.

BillWagner committed:
Update the index for the C# guide.

BillWagner committed:
fix OPS build errors.

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
```diff
@@ -0,0 +1,82 @@
+---
+title: C# Arrays | A tour of the C# language
```
I'm wondering if we want to establish a convention of which product names come after the pipe. I'd think we'd want just C# here.

BillWagner replied:
We want a convention. Is C# too broad? When these topics show up in search results, they will (in time) show up near results for more in depth coverage of the topic. Would developers be able to distinguish the overview from the in-depth topic? How much does that matter?

/cc @cartermp

cartermp replied:
That's an interesting problem.  I think with F# it's a bit easier since arrays are different (although they still have almost the same semantics), but it's still so many shared concepts.

I think it's best if we accept that there will be duplicated efforts here and cross-link.  The .NET arrays topic should link to the C# and F# tours.  Perhaps they should also link to the .NET arrays topic, even if it's mostly the same info.  That broadens search and gets us more hits.

mairaw replied:
Sure. But we're talking here how they'll be displayed by search engines @cartermp. This is what this metadata field is about.

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
````diff
@@ -0,0 +1,82 @@
+}
+```
+
+This example creates and operates on a ***single-dimensional array***. C# also supports ***multi-dimensional arrays***. The number of dimensions of an array type, also known as the ***rank*** of the array type, is one plus the number of commas written between the square brackets of the array type. The following example allocates a onesingle-dimensional, a two-dimensional, and a three-dimensional  ...
````
typo: onesingle

mairaw commented on docs/csharp/tour-of-csharp/attributes.md:
````diff
@@ -0,0 +1,55 @@
+}
+```
+
+All attribute classes derive from the System.Attribute base class provided by the standard library. Attributes can be applied by giving their name, along with any arguments, inside square brackets just before the associated declaration. If an attribute’s name ends in Attribute, that part of the name can be omitted when the attribute is referenced. For example, the `HelpAttribute` attribute can be ...
````
Don't you want to link to System.Attribute? You can just add the @ symbol.

mairaw commented on docs/csharp/tour-of-csharp/attributes.md:
````diff
@@ -0,0 +1,55 @@
+}
+```
+
+This example attaches a `HelpAttribute` to the `Widget` class ... and another `HelpAttribute` to the `Display` method in the class. The public constructors of an attribute class control the information that must be provided when the attribute is attached to a program entity. Additional information can be provided by referencing public read-write properties of the attribute class (such as the refe ...
````
Did you mean to have the ... here?

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,725 @@
+Point p1 = new Point(0, 0);
+Point p2 = new Point(10, 20);
+```
+The memory occupied by an object is automatically reclaimed when the object is no longer reachablein use. It is neither necessary nor possible to explicitly deallocate objects in C#.
````
typo: reachablein

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,725 @@
+}
+```
+
+A ***parameter array*** permits a variable number of arguments to be passed to a method. A parameter array is declared with the params modifier. Only the last parameter of a method can be a parameter array, and the type of a parameter array must be a single-dimensional array type. The Write and WriteLine methods of the System.Console class are good examples of parameter array usage. They are decl ...
````
should params be styled as inline code? Also, should we add links to the APIs you're mentioning here too?

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,725 @@
+
+An ***indexer*** is a member that enables objects to be indexed in the same way as an array. An indexer is declared like a property except that the name of the member is this followed by a parameter list written between the delimiters `[` and `]`. The parameters are available in the accessor(s) of the indexer. Similar to properties, indexers can be read-write, read-only, and write-only, and the a ...
+
+The `List` class declares a single read-write indexer that takes an `int` parameter. The indexer makes it possible to index `List` instances with `int` values. For example
```
For example is at least missing a colon

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,725 @@
+
+A constructor is declared like a method with no return type and the same name as the containing class. If a constructor declaration includes a static modifier, it declares a static constructor. Otherwise, it declares an instance constructor.
+
+Instance constructors can be overloaded, and can have optional parameters. For example, the `List<T>` class declares two instance constructors, one with no parameters and one that takes an int parameter. Instance constructors are invoked using the new operator. The following statements allocate two `List<string>` instances using the constructor of the `List` class with and without the optional ar ...
```
int parameter: int needs to be styled as inline code

mairaw commented on docs/csharp/tour-of-csharp/delegates.md:
```diff
@@ -0,0 +1,69 @@
+
+A delegate can reference either a static method (such as `Square` or `Math.Sin` in the previous example) or an instance method (such as `m.Multiply` in the previous example). A delegate that references an instance method also references a particular object, and when the instance method is invoked through the delegate, that object becomes this in the invocation.
+
+Delegates can also be created using anonymous functions, which are "in-line methods" that are created on the fly. Anonymous functions can see the local variables of the sourrounding methods. Thus, the multiplier example above can be written more easily without using a Multiplier class:
```
typo: sourrounding

mairaw commented on docs/csharp/tour-of-csharp/enums.md:
```diff
@@ -0,0 +1,82 @@
+	
+# Enums
+
+An ***enum type*** is a distinct value type with a set of named constants. The following example declares and uses an enum type named `Color` with three constant values, `Red`, `Green`, and `Blue`.
```
I feel like the introduction of the example is kind of abrupt. You have just a short summary and immediately comes the following example.

mairaw commented on docs/csharp/tour-of-csharp/structs.md:
````diff
@@ -0,0 +1,68 @@
+Console.WriteLine(b.x);
+```
+
+If `Point` is a class, the output is 20 because a and b reference the same object. If Point is a struct, the output is 10 because the assignment of a to b creates a copy of the value, and this copy is unaffected by the subsequent assignment to a.x.
````
I'd style a, b. a.x as inline code. I'm just thinking when we go to localize these docs, it's better if the non-localizable segments are locked like variable names, etc.

mairaw commented on docs/csharp/features.md:
```diff
@@ -1,27 +0,0 @@
----
```
Now that the content is live, do we want to set up redirection for topics we're deleting or renaming? There's a way to do it, but it requires you to keep the old file there (you just need to add some metadata to it) :cry: 
I haven't done myself but I've seen the instructions on internal guidance sites

BillWagner replied:
If this document had content, I'd agree with setting up the redirect. I think we can avoid that because it was a placeholder document.

mairaw replied:
Sounds good!

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,725 @@
+
+### Virtual, override, and abstract methods
+
+When an instance method declaration includes a `virtual` modifier, the method is said to be a ***virtual method***. When no virtual modifier is present, the method is said to be a ***non-virtual method***.
```
you use nonvirtual and non-virtual in this topic. I've consulted Microsoft's style guide and we should only hyphenate words beginning with non if necessary to avoid confusion.

mairaw commented:
I did a very quick edit pass on the topics. I can review them more carefully later on.

BillWagner committed:
First pass update from reviews.

BillWagner committed:
Full proof read and style check.

BillWagner commented:
@mairaw I read all your feedback, and did a sweep of all the content to check for similar errors. I'd like another review, and then hopefully it's ready to go.

mairaw commented:
Sounds good.

mairaw commented on docs/csharp/index.md:
```diff
@@ -14,49 +14,22 @@ ms.assetid: 52db8280-0e53-40cf-858b-e8eef3997dea
+
+If you prefer reading overview information first, try our [tour of the C# language](tour-of-csharp/index.md). It explains the concepts of the C# language. After reading this, you'll have a basic understanding of the language, and be ready to try the tutorials, or build something on your own.
+
+## For Developers new to C#
```
you'll need to add the # after c# to have this rendering correctly

mairaw replied:
nit: also on new topics we should try to follow our guidelines for sentence-case on headings and titles. less things to fix later on :smile:

mairaw commented on docs/csharp/index.md:
```diff
@@ -14,49 +14,22 @@ ms.assetid: 52db8280-0e53-40cf-858b-e8eef3997dea
-* [🔧 Syntax Reference](syntax.md)
+The C# guide provides a wealth of information about the C# language. This site has many different audiences. Depending on your experience with programming, or with the C# language and .NET, you may wish to explore different sections of this guide.
+
+## For new developers
```
should we call this section For brand-new developers? Just to make a clearer distinction with the next section?

mairaw commented on docs/csharp/index.md:
```diff
@@ -14,49 +14,22 @@ ms.assetid: 52db8280-0e53-40cf-858b-e8eef3997dea
+
+You can also browse the [tutorials](tutorials/index.md) to try basic C# programs.
+
+Or, check out the [C# Concepts](concepts.md) to learn the features of the C# langauge in depth.
```
typo: langauge

mairaw commented on docs/csharp/index.md:
```diff
@@ -14,49 +14,22 @@ ms.assetid: 52db8280-0e53-40cf-858b-e8eef3997dea
+
+## For Developers new to C#
+
+If you've done development before, but are new to C#, read the [tour of the C# language](tour-of-csharp/index.md). It will provide with a starting point, and you can use it to contrast C# with other languages you've used.
```
It will provide with a starting point reads a bit funny. It will provide you with a starting point or It will provide a starting point perhaps?

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
```diff
@@ -0,0 +1,86 @@
+
+An ***array*** is a data structure that contains a number of variables that are accessed through computed indices. The variables contained in an array, also called the ***elements*** of the array, are all of the same type, and this type is called the ***element type*** of the array.
+
+Array types are reference types, and the declaration of an array variable simply sets aside space for a reference to an array instance. Actual array instances are created dynamically at run-time using the new operator. The new operation specifies the ***length*** of the new array instance, which is then fixed for the lifetime of the instance. The indices of the elements of an array range from `0` ...
```
from our old guidelines, run-time here should be not hyphened and spelled like this: run time.
Use run time to indicate the period of time during which a program is running. 
Use runtime as an abbreviation for common language runtime or, in a more generic sense, to indicate an active program during run time. 
Use run-time as an adjective.

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
````diff
@@ -0,0 +1,86 @@
+```
+
+The `a1` array contains 10 elements, the `a2` array contains 50 (10 × 5) elements, and the `a3` array contains 100 (10 × 5 × 2) elements.
+The element type of an array can be any type, including an array type. An array with elements of an array type is sometimes called a ***jagged array*** because the lengths of the element arrays do not all have to be the same. The following example allocates an array of arrays of int:
````
arrays of int: shouldn't int be formatted as inline code here?

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
````diff
@@ -0,0 +1,86 @@
+a[2] = new int[20];
+```
+
+The first line creates an array with three elements, each of type int[] and each with an initial value of null. The subsequent lines then initialize the three elements with references to individual array instances of varying lengths.
````
same here for int[] and null

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,728 @@
+
+***Classes*** are the most fundamental of C#’s types. A class is a data structure that combines state (fields) and actions (methods and other function members) in a single unit. A class provides a definition for dynamically created ***instances*** of the class, also known as ***objects***. Classes support ***inheritance*** and ***polymorphism***, mechanisms whereby ***derived classes*** can exten ...
+
+New classes are created using class declarations. A class declaration starts with a header that specifies the attributes and modifiers of the class, the name of the class, the base class (if given), and the interfaces implemented by the class. The header is followed by the class body, which consists of a list of member declarations written between the delimiters `{` and `}`.
```
I noticed it here but maybe there are other sections that could benefit from the same strategy. Wouldn't be good to link to existing topics when you mention them? For example, in this sentence: "with a header that specifies the attributes" wouldn't be good to link to the attributes topic?

BillWagner replied:
Yes. I want to defer that task until after this PR and PR #782 get merged. Then, I'll add links in existing content to match the new locations.

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,728 @@
+* `public`
+	- Access not limited
+* `protected`
+	- Access limited to this class or classes derived from this class
```
since there's a not a class or program in discussion, should these descriptions be more generic?

BillWagner replied:
I looked at this, and I'd rather leave the text as is. This topic introduces 'class', so member of a class makes sense in that context. If this section is read in order, 'struct' has not been introduced, so I'd rather not mention it. Program text has been introduced. It's the term used in 'program-structure' for how programs are organized.

Any more generic text starts to introduce more terms for the reader.

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,728 @@
+}
+```
+
+A class inherits the members of its base class. Inheritance means that a class implicitly contains all members of its base class, except for the instance and static constructors, and the finalizers of the base class. A derived class can add new members to those it inherits, but it cannot remove the definition of an inherited member. In the previous example, `Point3D` inherits the `x` and `y` fiel ...
````
Not sure if we need to talk about access to private members here. They're inherited but you cannot access them unless they're in a nested class.

BillWagner replied:
This being the overview, I want to keep it as simple as possible. I'm OK with omissions, as long as the topic is correct.

mairaw replied:
Yep, topic is correct.

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,728 @@
+
+A ***method*** is a member that implements a computation or action that can be performed by an object or class. ***Static methods*** are accessed through the class. ***Instance methods*** are accessed through instances of the class.
+
+Methods have a (possibly empty) list of ***parameters***, which represent values or variable references passed to the method, and a ***return type***, which specifies the type of the value computed and returned by the method. A method’s return type is void if it does not return a value.
```
wouldn't be easier to say "Methods can have a list of **parameters**? and why are you using three asterisks instead of two?

mairaw replied:
also style void as inline code or this will get localized in the future :smile:

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,728 @@
+
+Value parameters can be optional, by specifying a default value so that corresponding arguments can be omitted.
+
+A ***reference parameter*** is used for both input and output parameter passing. The argument passed for a reference parameter must be a variable, and during execution of the method, the reference parameter represents the same storage location as the argument variable. A reference parameter is declared with the ref modifier. The following example shows the use of ref parameters.
```
same here, style ref as inline code

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,728 @@
+}
+```
+
+An ***output parameter*** is used for output parameter passing. An output parameter is similar to a reference parameter except that the initial value of the caller-provided argument is unimportant. An output parameter is declared with the out modifier. The following example shows the use of out parameters.
````
Ditto for out

dend replied:
For snippets, can we refer them from `/samples/snippets/csharp` instead of embedding them in the doc?

mairaw replied:
I've suggested this to @BillWagner too. We could just merge this and then later add the snippets if we need to get this out quickly. Also @dend, have we settled on the /samples/snippets folder? Is that configurable?

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,728 @@
+
+A method declared without a static modifier is an ***instance method***. An instance method operates on a specific instance and can access both static and instance members. The instance on which an instance method was invoked can be explicitly accessed as `this`. It is an error to refer to `this` in a static method.
+
+The following Entity class has both static and instance members.
```
Ditto for Entity

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,728 @@
+	new Constant(3));
+```
+
+The `Evaluate` method of an `Expression` instance is invoked to evaluate the given expression and produce a `double` value. The method takes as an argument a `Hashtable` that contains variable names (as keys of the entries) and values (as values of the entries). The `Evaluate` method is a `virtual abstract` method, meaning that non-abstract derived classes must override it to provide an actual im ...
````
might want to link to the Hashtable api ref

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
````diff
@@ -0,0 +1,728 @@
+		Console.WriteLine(a == b);  // Outputs "False"
+	}
+}
+```
````
Is the List you're using here the one you showed before? I was wondering if we should start adding snippets for these examples so that they go through CI and so forth. (could be a later goal)
I'm also asking this because I couldn't use the classes as-is. For example, the List<T> code is missing using System or overriding GetHashCode and I get an exception when I run this code (it's in an infinite loop because of the == operator perhaps and the fact that the Equals method calls == again).

BillWagner replied:
Great catch. I updated the code for Equals() to fix that. 

And yes, this should go through a CI build. I'll do that next commit. (And I'll fix the GetHashCode() warning).

mairaw commented on docs/csharp/tour-of-csharp/delegates.md:
```diff
@@ -0,0 +1,73 @@
+
+A ***delegate type*** represents references to methods with a particular parameter list and return type. Delegates make it possible to treat methods as entities that can be assigned to variables and passed as parameters. Delegates are similar to the concept of function pointers found in some other languages, but unlike function pointers, delegates are object-oriented and type-safe.
+
+The following example declares and uses a delegate type named Function.
```
Function missing formatting

mairaw commented on docs/csharp/tour-of-csharp/delegates.md:
```diff
@@ -0,0 +1,73 @@
+
+A delegate can reference either a static method (such as `Square` or `Math.Sin` in the previous example) or an instance method (such as `m.Multiply` in the previous example). A delegate that references an instance method also references a particular object, and when the instance method is invoked through the delegate, that object becomes this in the invocation.
+
+Delegates can also be created using anonymous functions, which are "in-line methods" that are created on the fly. Anonymous functions can see the local variables of the surrounding methods. Thus, the multiplier example above can be written more easily without using a Multiplier class:
```
in-line -> inline

mairaw commented on docs/csharp/tour-of-csharp/enums.md:
```diff
@@ -0,0 +1,88 @@
+
+As shown by the previous example, an `enum` member declaration can include a constant expression that specifies the value of the member. The constant value for each `enum` member must be in the range of the underlying type of the `enum`. When an `enum` member declaration does not explicitly specify a value, the member is given the value zero (if it is the first member in the `enum` type) or the v ...
+
+`Enum` values can be converted to integral values and vice versa using type casts. For example
```
nit: For example is missing punctuation

mairaw commented on docs/csharp/tour-of-csharp/index.md:
```diff
@@ -0,0 +1,86 @@
+
+# A Tour of the C# Language  
+
+C# (pronounced “See Sharp”) is a simple, modern, object-oriented, and type-safe programming language. C# has its roots in the C family of languages and will be immediately to C, C++, Java, and JavaScript programmers.
```
The pronunciation bit will be hard to translate :smile:

mairaw replied:
incomplete sentence: "will be immediately to C, C++"

mairaw commented on docs/csharp/tour-of-csharp/index.md:
````diff
@@ -0,0 +1,86 @@
+C# source files typically have the file extension `.cs`. Assuming that the “Hello, World” program is stored in the file `hello.cs`, the program might be compiled using the command line:
+
+```
+csc hello.cs
````
newbie question: isn't csc for framework compilation? if so, do we need to call that out?

BillWagner replied:
Done.

I've got an email out to @blackdwarf to figure out how to make it work on core.

BillWagner replied:
On core, we'll use the `dotnet` CLI, and I've put a note in for that.

mairaw commented on docs/csharp/tour-of-csharp/interfaces.md:
````diff
@@ -0,0 +1,88 @@
+IDataBound dataBound = (IDataBound)obj;
+```
+
+In the previous `EditBox` class, the `Paint` method from the `IControl` interface and the `Bind` method from the `IDataBound` interface are implemented using public members. C# also supports explicit ***interface member implementations***, using which the class or struct can avoid making the members public. An explicit interface member implementation is written using the fully qualified interface ...
````
this part reads a bit funny: using which the class

mairaw commented on docs/csharp/tour-of-csharp/program-structure.md:
````diff
@@ -0,0 +1,106 @@
+}
+```
+
+declares a class named `Stack` in a namespace called `Acme.Collections`. The fully qualified name of this class is `Acme.Collections.Stack`. The class contains several members: a field named `top`, two methods named `Push` and `Pop`, and a nested class named `Entry`. The `Entry` class further contains three members: a field named `next`, a field named `data`, and a constructor. Assuming that the  ...
````
I find it odd to break the first sentence as this. Maybe move that first part above the example?

BillWagner committed:
Respond to feedback

Also, update .gitignore because the api refs are now put in a different
directory.

BillWagner commented:
Just added the WIP label. I responded to all content feedback. I'm going to move the code snippets to sub-directories so they are all handled by the CI build system.

mairaw commented:
Question: when you say putting snippets in sub-directories, what do you mean? I thought we had a separate folder for snippets, no?

BillWagner commented:
@mairaw Yes, I mean putting the snippets in the snippets folder. 
There are multiple samples under this PR, so it will be in a few sub-dirs. under the snippets folder.

mairaw commented:
Ah got it!

BillWagner committed:
using snippets up through classes topic

BillWagner committed:
done through enums

BillWagner committed:
finish converting to Snippets

BillWagner committed:
Add a note regarding xplat-tools

BillWagner committed:
add a note about xplat tools

BillWagner commented:
Removing the WIP label. This should be all set to merge.

BillWagner committed:
update xplat note.

mairaw commented on docs/csharp/tour-of-csharp/classes-and-objects.md:
```diff
@@ -0,0 +1,297 @@
+The `using` statement provides a better approach to object destruction.
+
+>[!div class="step-by-step"]
+[Pre](statements.md)
```
nit: I know this is how it's on the template but I was wondering if we should change it to something like « Previous and Next » now that I see how this text renders on stage

BillWagner replied:
I did change to "Previous", which I think reads better.

mairaw commented on docs/csharp/tour-of-csharp/arrays.md:
```diff
@@ -0,0 +1,50 @@
+
+# Arrays
+
+An ***array*** is a data structure that contains a number of variables that are accessed through computed indices. The variables contained in an array, also called the ***elements*** of the array, are all of the same type, and this type is called the ***element type*** of the array.
```
nit: maybe I've asked this before but why *** instead of **?

BillWagner replied:
I use ***the bold italic*** when I introduce a new term.

mairaw replied:
Ah, got it! Learning new things every day. :smile:

mairaw commented on docs/csharp/tour-of-csharp/attributes.md:
```diff
@@ -0,0 +1,32 @@
+When a particular attribute is requested through reflection, the constructor for the attribute class is invoked with the information provided in the program source, and the resulting attribute instance is returned. If additional information was provided through properties, those properties are set to the given values before the attribute instance is returned.
+
+>[!div class="step-by-step"]
+[Pre](attributes.md)
```
this is not going to the previous topic (it's actually linking it to itself) and it's missing the next button to go to the what's new topic apparently.

BillWagner replied:
I updated the Previous Link.

When I update the TOC, what about an "Up" button to back to the index? The "Next" file in the TOC is unrelated.

mairaw replied:
Just fix the Previous link then, I think.

mairaw commented:
LGTM modulo the minor comments I left

BillWagner committed:
Final feedback.
